src/combinatorics/hales_jewett.lean

```diff
@@ -0,0 +1,299 @@
+`line m n` defining the same function. -/
+structure line (m n : Type) : Type :=
+(idx_fun : n → option m)
+(proper : ∃ i, idx_fun i = none)
+
+namespace line
+
+/- This lets us treat a line `l : line m n` as a function `m → (n → m)`. -/
+instance (m n) : has_coe_to_fun (line m n) :=
+⟨λ _, m → n → m, λ l x i, (l.idx_fun i).get_or_else x⟩
+
+/-- The type of lines that are only one color except possibly at their endpoints. -/
+@[nolint has_inhabited_instance]
+structure almost_mono {m n k : Type} (C : (n → option m) → k) :=
+(line : line (option m) n)
+(color : k)
+(has_color : ∀ x : m, C (line (some x)) = color)
+
+/-- The type of families of lines such that
+- each line is only one color except possibly at its endpoint
+- the lines all have the same endpoint
+- the colors of the lines are distinct.
+Used in the proof `exists_mono_in_high_dimension`. -/
+structure color_focused {m n k} (C : (n → option m) → k) :=
+(lines : multiset (almost_mono C))
+(focus : n → option m)
+(is_focused : ∀ p : almost_mono C, p ∈ lines → p.line none = focus)
+(distinct_colors : (lines.map almost_mono.color).nodup)
+
+instance {m n k} (C : (n → option m) → k) : inhabited (color_focused C) :=
+⟨⟨0, λ _, none, λ _, false.elim, multiset.nodup_zero⟩⟩
+
+/-- The diagonal line. It is the identity at every coordinate. -/
+def diagonal (m n) [nonempty n] : line m n :=
+{ idx_fun := λ _, none,
+  proper  := ⟨classical.arbitrary n, rfl⟩ }
+
+instance (m n) [nonempty n] : inhabited (line m n) := ⟨diagonal m n⟩
+
+/-- A line is monochromatic if all its points are the same color. -/
+def is_mono {m n k : Type} (C : (n → m) → k) (l : line m n) : Prop :=
+∃ c, ∀ x, C (l x) = c
+
+/-- A function `m → m'` determines a function `line m n → line m' n`. For a coordinate `i`,
+`l.map f` is the identity at `i` if `l` is, and constantly `f y` if `l` is constantly `y` at `i`. -/
+def map {m m' n} (f : m → m') (l : line m n) : line m' n :=
+{ idx_fun := λ i, (l.idx_fun i).map f,
+  proper  := ⟨l.proper.some, by rw [l.proper.some_spec, option.map_none'] ⟩ }
+
+/-- A point in `n → m` and a line in `n' → m` determine a line in `n ⊕ n' → m`. -/
+def vertical {m n n'} (v : n → m) (l : line m n') : line m (n ⊕ n') :=
+{ idx_fun := sum.elim (some ∘ v) l.idx_fun,
+  proper  := ⟨sum.inr l.proper.some, l.proper.some_spec⟩ }
+
+/-- A line in `n → m` and a point in `n' → m` determine a line in `n ⊕ n' → m`. -/
+def horizontal {m n n'} (l : line m n) (v : n' → m) : line m (n ⊕ n') :=
+{ idx_fun := sum.elim l.idx_fun (some ∘ v),
+  proper  := ⟨sum.inl l.proper.some, l.proper.some_spec⟩ }
+
+/-- One line in `n → m` and one in `n' → m` together determine a line in `n ⊕ n' → m`. -/
+def prod {m n n'} (l : line m n) (l' : line m n') : line m (n ⊕ n') :=
+{ idx_fun := sum.elim l.idx_fun l'.idx_fun,
+  proper  := ⟨sum.inl l.proper.some, l.proper.some_spec⟩ }
+
+lemma apply {m n} (l : line m n) (x : m) : l x = λ i, (l.idx_fun i).get_or_else x := rfl
```

Comment:
This lemma lets you unfold the definition of the coe to fun instance on `line`s, perhaps it could have a name reflecting that? `line_apply` perhaps?

Reply:
The full name is `combinatorics.line.apply` (and the `line` namespace should not be opened), is this ok? I'm happy to change it

Reply:
That's fine!